esmvalcore/_recipe/from_datasets.py

```diff
@@ -0,0 +1,327 @@
+                facets['ensemble'] = ensemble
+                result.append(facets)
+    return result
+
+
+def _group_ensemble_names(ensemble_names: Iterable[str]) -> list[str]:
+    """Group ensemble names.
+
+    Examples
+    --------
+    ensemble_names=[
+        'r1i1p1',
+        'r2i1p1',
+        'r3i1p1',
+        'r1i1p2',
+    ]
+    will return [
+        'r(1:3)i1p1',
+        'r1i1p2',
+    ].
+    """
+    ensemble_tuples = [
+        tuple(int(i) for i in re.findall(r'\d+', ens))
+        for ens in ensemble_names
+    ]
+
+    ensemble_ranges = _create_ensemble_ranges(ensemble_tuples)
+
+    groups = []
+    for ensemble_range in ensemble_ranges:
+        txt = ''
+        for name, value in zip('ripf', ensemble_range):
+            txt += name
+            if value[0] == value[1]:
+                txt += f"{value[0]}"
+            else:
+                txt += f"({value[0]}:{value[1]})"
+        groups.append(txt)
+
+    return groups
+
+
+def _create_ensemble_ranges(
+    ensembles: Sequence[tuple[int,
+                              ...]], ) -> list[tuple[tuple[int, int], ...]]:
+    """Create ranges from tuples.
+
+    Examples
+    --------
+    Input ensemble member tuple (1, 1, 1) represents 'r1i1p1'.
+    The input tuples will be converted to ranges, for example
+    ensembles=[
+        (1, 1, 1),
+        (2, 1, 1),
+        (3, 1, 1),
+        (1, 1, 2),
+    ]
+    will return [
+        ((1, 3), (1, 1), (1, 1)),
+        ((1, 1), (1, 1), (2, 2)),
+    ].
+    """
+
+    def order(i, ens):
+        prefix, suffix = ens[:i], ens[i + 1:]
+        return (prefix, suffix, ens[i])
+
+    def grouper(i, ens):
+        prefix, suffix = ens[:i], ens[i + 1:]
+        return (prefix, suffix)
+
+    for i in range(len(ensembles[0])):
+        grouped_ensembles = []
+        ensembles = sorted(ensembles, key=partial(order, i))
+        for (prefix,
+             suffix), ibunch in itertools.groupby(ensembles,
+                                                  key=partial(grouper, i)):
+            bunch = list(ibunch)
+            prev = bunch[0][i]
+            groups = [[prev]]
+            for ensemble in bunch[1:]:
+                if ensemble[i] == prev + 1:
+                    prev += 1
+                else:
+                    groups[-1].append(prev)
+                    prev = ensemble[i]
+                    groups.append([prev])
+            groups[-1].append(prev)
+            result = []
+            for group in groups:
+                item = prefix + (tuple(group), ) + suffix
+                result.append(item)
+            grouped_ensembles.extend(result)
+
+        ensembles = grouped_ensembles
+
+    return sorted(ensembles)  # type: ignore
+
+
+def datasets_to_recipe(
+    datasets: Iterable[Dataset],
+    recipe: dict | None = None,
+) -> dict:
+    """Create or update a recipe from datasets.
+
+    Parameters
+    ----------
+    datasets
+        Datasets to use in the recipe.
+    recipe
```

Comment:
mention data type (`dict`)

Reply:
The data types are shown in the function signature based on the type hints: https://esmvaltool--1877.org.readthedocs.build/projects/ESMValCore/en/1877/api/esmvalcore.dataset.html#esmvalcore.dataset.datasets_to_recipe

It seems duplicate to then repeat them in the docstrings.

Reply:
The updated documentation, with an improved description of what the `recipe` argument and the return value are, is now available [here](https://esmvaltool--1877.org.readthedocs.build/projects/ESMValCore/en/1877/api/esmvalcore.dataset.html#esmvalcore.dataset.datasets_to_recipe).

Reply:
perfect